File: 0680/solution.py
class Solution:
    def validPalindrome(self, s: str) -> bool:

        def pal(l, r):
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True

        left, right = 0, len(s) - 1
        while left < right:
            if s[left] != s[right]:
                if pal(left, right - 1) or pal(left + 1, right):
                    return True
                return False
            left += 1
            right -= 1
        return True

        # time O(n)
        # space O(1)

File: 0680/solution.rs
impl Solution {
    pub fn valid_palindrome(s: String) -> bool {
        let bytes = s.as_bytes();
        Self::palindrome_helper(bytes, 0, bytes.len() - 1, 0)
    }

    fn palindrome_helper(s: &[u8], mut left: usize, mut right: usize, errors: usize) -> bool {
        while left < right {
            if s[left] != s[right] {
                if errors == 1 {
                    return false;
                }
                return Self::palindrome_helper(s, left + 1, right, errors + 1)
                    || Self::palindrome_helper(s, left, right - 1, errors + 1);
            }
            left += 1;
            right -= 1;
        }
        true
    }
}
